### PR TITLE
Add support for getting a &str from a GValue without copying

### PR DESCRIPTION
**Pull request by sdroege**
Needs some discussion if the lifetime parameter is good or not here. For G_TYPE_STRING the copying or not might not be that important, but I'll also have to implement this for more complex types in the future where copying is expensive.

Without the added lifetime parameter, the following would compile just fine (while it shouldn't):
```rust
    let s = {
        let v = Value::from("test");
        assert_eq!(v.get::<&str>(), Some("test"));
        v.get::<&str>()
    };

    assert_eq!(s, Some("test"));
```

(Note that using get::<String>() instead above still works, as it should)

### CODE REVIEW

**Comment by sdroege** [comment]
If this is accepted, I would also add something that can convert from a TypedValue<T: ToOwned> to a TypedValue<T::Owned> and back to allow easy switching between e.g. TypedValue<String> and TypedValue<&str>.

**Comment by sdroege** [comment]
@GuillaumeGomez wrote on IRC that this seems ok to him. So next step: 1) fixing all implementations so the tests succeed, 2) implement something to convert between e.g. TypedValue<String> and TypedValue<&str> easily.

Any suggestions for the API for 2)? Or maybe just always work with TypedValue<String> but have a get_ref() on it if FromValueOptional is implemented for the types involved, and one implements Borrow for the other?

**Comment by EPashkin** [comment]
For first part only gdk need fix.
We don't have other implementation currently. and we only need it in gdk as I unable to define implementations for some `BoxedValue` parallel with `IsA<Object>` implementations
👍 for merge this ignoring build error.

**Comment by sdroege** [comment]
For 2) I don't see any useful way of doing that generically, because ToOwned/Borrowed work with "str" as type instead of "&str". We could define our own mapping traits here, which seems like quite some overhead, or just implement From<_> for the conversion between TypedValue<&str> and TypedValue<String> in both directions.

@GuillaumeGomez what do you think?

File: src/value.rs [hunk 1]
```diff
@@ -427,12 +440,23 @@ pub trait SetValue: StaticType {
...
 }
 
+impl<'a> FromValueOptional<'a> for &'a str {
+    unsafe fn from_value_optional(value: &'a Value) -> Option<Self> {
+        let cstr = gobject_ffi::g_value_get_string(value.to_glib_none().0);
+        if cstr.is_null() {
+            return None;
+        }
+
+        CStr::from_ptr(cstr).to_str().ok()
```
Comment (GuillaumeGomez) [inline src/value.rs]: Why not just an `else` condition and remove the `return` statement above?

**Comment by GuillaumeGomez** [comment]
Just a small nit/suggestion but otherwise it's all good.

File: src/value.rs (hunk 1 above)
Comment (sdroege) [inline src/value.rs]: Done :) Just a matter of taste I guess, I prefer to have the "error" code paths outside the normal code path.

**Comment by GuillaumeGomez** [comment]
@EPashkin: Ok for me, tell me when it's ok for you as well so I can merge.

**Comment by EPashkin** [comment]
@GuillaumeGomez yes, IMHO in can be merged

**Comment by GuillaumeGomez** [comment]
Great! Thanks!

**Pull request merged**